src/accelerate/data_loader.py

```diff
@@ -870,6 +870,11 @@ def prepare_data_loader(
         sampler = getattr(dataloader.sampler, "sampler", None)
     else:
         sampler = getattr(dataloader.batch_sampler, "sampler", None)
+    if isinstance(dataloader.sampler, RandomSampler) and state.distributed_type == DistributedType.XLA:
+        # isinstance(dataloader.sampler, RandomSampler) indicates the original dataloader has `shuffle` enabled.
+        generator = torch.Generator().manual_seed(42)
+        dataloader.generator = generator
+        dataloader.sampler.generator = generator
```

Comment:
Is this sort of logic going to have us run into issues when `use_seedable_sampler=True`? 

Also not sure setting `dataloader.generator` truly does anything, as we recreate a new dataloader anyways later?

We should be adjusting `dataloader.sampler.generator`, which *really* should be `sampler.generator`

Reply:
I'm actually not sure how this fixes anything, as we again fully recreate the dataloader, so setting any attribute inside `dataloader` is moot

Reply:
Same reasoning, that doesn't do anything. We need to adjust `sampler`, we should never be modifying the dataloader directly

Reply:
> Also not sure setting dataloader.generator truly does anything, as we recreate a new dataloader anyways later?
> I'm actually not sure how this fixes anything, as we again fully recreate the dataloader, so setting any attribute inside dataloader is moot

Actually, setting dataloader.generator has some impact on how we recreate a new dataloader: here https://github.com/huggingface/accelerate/blob/60bfdaa9349e3e1a373863a8c5f20baa89cfd000/src/accelerate/data_loader.py#L921 we extract the `dataloader.generator` and then pass it in when we recreate a new dataloader here https://github.com/huggingface/accelerate/blob/60bfdaa9349e3e1a373863a8c5f20baa89cfd000/src/accelerate/data_loader.py#L961. I also set `dataloader.batch_sampler.sampler` here because `batch_sampler` is used to create the `new_batch_sampler` https://github.com/huggingface/accelerate/blob/60bfdaa9349e3e1a373863a8c5f20baa89cfd000/src/accelerate/data_loader.py#L901.

It is a tricky function to get it work correctly as it reused the dataloader that users pass in and reconstruct a new one. Initially we figured it out how to make it work by changing [line](https://github.com/huggingface/accelerate/blob/60bfdaa9349e3e1a373863a8c5f20baa89cfd000/src/accelerate/test_utils/scripts/test_script.py#L276) to `dl = DataLoader(range(length), batch_size=8, shuffle=True, generator=torch.Generator().manual_seed(42))`, by adding `generator`, as it is mandatory for TPU v2/v3 as in [here](https://github.com/pytorch/xla/blob/5c4ed89bc7bb3f7757ca403be887b25d1bdedfe1/test/pjrt/test_train_hf_transformer.py#L41). But I don't want the accelerate user to fail silently if they don't use `generator` when creating a dataloader with shuffling. That's why I decided to change the non-test code. I basically set things similar to what [DataLoader](https://github.com/pytorch/xla/blob/5c4ed89bc7bb3f7757ca403be887b25d1bdedfe1/test/pjrt/test_train_hf_transformer.py#L41) does when `shuffle` is True.

Reply:
Fantastic, thanks for the explanation!